src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,289 @@
+                textBuffer, _languageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+            // If there is no matching LSP server registered the client will return null here - e.g. wrong content type on the buffer, the
+            // server totally failed to start, server doesn't support the right capabilities. For C# we might know it's a bug if we get a null
+            // response here, but we don't know that in general for all languages.
+            // see "Microsoft.CodeAnalysis.Editor.Implementation.LanguageClient.AlwaysActivateInProcLanguageClient" for the list of content types we register for.
+            // At this time the expected list is C#, Visual Basic, and F#
+            if (response is null)
+            {
+                return null;
+            }
+
+            var responseBody = response.Value.response.ToObject<DocumentSymbol[]>();
+            // It would be a bug in the LSP server implementation if we get back a null result here.
+            Assumes.NotNull(responseBody);
+
+            var model = DocumentOutlineHelper.CreateDocumentSymbolDataModel(responseBody, response.Value.snapshot);
+
+            _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, CaretPositionOfNodeToSelect: null, ExpansionOption: null, DataUpdated: true));
+
+            return model;
+        }
+
+        private void EnqueueFilter(string? newText)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(newText, CaretPositionOfNodeToSelect: null, ExpansionOption: null, DataUpdated: false));
+
+        public void EnqueueSelectTreeNode(CaretPosition caretPoint)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, caretPoint, ExpansionOption: null, DataUpdated: false));
+
+        public void EnqueueExpandOrCollapse(ExpansionOption option)
+            => _updateViewModelStateQueue.AddWork(new ViewModelStateDataChange(SearchText, CaretPositionOfNodeToSelect: null, option, DataUpdated: false));
+
+        private async ValueTask UpdateViewModelStateAsync(ImmutableSegmentedList<ViewModelStateDataChange> viewModelStateData, CancellationToken cancellationToken)
+        {
+            _threadingContext.ThrowIfNotOnBackgroundThread();
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var model = await _documentSymbolQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            var searchText = viewModelStateData.SelectLastNonNullOrDefault(static x => x.SearchText);
+            var position = viewModelStateData.SelectLastNonNullOrDefault(static x => x.CaretPositionOfNodeToSelect);
+            var expansion = viewModelStateData.SelectLastNonNullOrDefault(static x => x.ExpansionOption);
+            var dataUpdated = viewModelStateData.Any(static x => x.DataUpdated);
+
+            // These updates always require a valid model to perform
```

Comment:
nit: your life would likely be easier if we always had a mdoel (even one representing the empty state).  but up to you if you want to go that route.